presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java

```diff
@@ -6963,4 +6963,26 @@ public void testLeftJoinNullFilterToSemiJoin()
         sql = "with t as (select orderkey from orders where custkey%5=1) select l.orderkey, l.partkey, l.quantity from lineitem l left join t on l.orderkey = t.orderkey and l.suppkey%5=1 where t.orderkey is null";
         assertQuery(enableOptimization, sql);
     }
+
+    @Test
+    public void testCaseToMapOptimization()
+    {
+        assertNotEquals(computeActual("EXPLAIN(TYPE DISTRIBUTED) SELECT orderkey, CASE  WHEN orderstatus||'-'|| comment ='O' THEN 'a' " +
+                "when  orderstatus||'-'|| comment='F' THEN 'b' else 'other' END, " +
+                "sum(case when  orderstatus||'-'|| comment='F' THEN  totalprice END)" +
+                "FROM orders group by 1,2").getOnlyValue().toString().indexOf("element_at"), -1);
+
+        assertNotEquals(computeActual("EXPLAIN(TYPE DISTRIBUTED) SELECT orderkey, CASE  WHEN orderstatus in ('a', 'f', 'b') THEN 'a' " +
+                "when  orderstatus='F' THEN 'b' else 'other' END, " +
+                "sum(case when  orderstatus||'-'|| comment='F' THEN  totalprice END)" +
+                "FROM orders group by 1,2").getOnlyValue().toString().indexOf("element_at"), -1);
+
+        // Should not be applied below
+        assertEquals(computeActual("EXPLAIN(TYPE DISTRIBUTED) SELECT orderkey, CASE  WHEN orderstatus in ('a', 'f', 'b') THEN comment||'a' " +
+                "when  orderstatus='F' THEN comment||'b' else 'other' END, " +
+                "sum(case when  orderstatus||'-'|| comment='F' THEN  totalprice END)" +
+                "FROM orders group by 1,2").getOnlyValue().toString().indexOf("element_at"), -1);
+
+        assertQuery("select case x when 1 then 1 when 2 then 2 when 1 then 3 end from (select 1 as x) t", "select 1");
+    }
```

Comment:
Can we add a few more cases? for example
select x, case x when 1 then 1 when 2 then 2 else 3 end from (select x from (values 1, 2, 3, 4) t(x))

select x, case when x=1 then 1 when x=2 then 2 else 3 end from (select x from (values 1, 2, 3, 4) t(x))

select x, case when x=1 then 1 when x in array[2, 3] then 2 else 3 end from (select x from (values 1, 2, 3, 4) t(x))

Reply:
Done